pennylane/transforms/decompose.py

```diff
@@ -39,27 +40,72 @@ def _operator_decomposition_gen(
     acceptance_function: Callable[[qml.operation.Operator], bool],
     max_expansion: Optional[int] = None,
     current_depth=0,
+    graph=None,
 ) -> Generator[qml.operation.Operator, None, None]:
     """A generator that yields the next operation that is accepted."""
 
     max_depth_reached = False
+    decomp = []
 
     if max_expansion is not None and max_expansion <= current_depth:
         max_depth_reached = True
 
     if acceptance_function(op) or max_depth_reached:
         yield op
+    elif graph is not None and graph.is_solved_for(op):
+        op_rule = graph.decomposition(op)
+        with qml.queuing.AnnotatedQueue() as decomposed_ops:
+            op_rule(*op.parameters, wires=op.wires, **op.hyperparameters)
+        decomp = decomposed_ops.queue
+        current_depth += 1
     else:
         decomp = op.decomposition()
         current_depth += 1
 
-        for sub_op in decomp:
-            yield from _operator_decomposition_gen(
-                sub_op,
-                acceptance_function,
-                max_expansion=max_expansion,
-                current_depth=current_depth,
-            )
+    for sub_op in decomp:
+        yield from _operator_decomposition_gen(
+            sub_op,
+            acceptance_function,
+            max_expansion=max_expansion,
+            current_depth=current_depth,
+            graph=graph,
+        )
+
+
+def _graph_decomps_kwargs_checks(fixed_decomps, alt_decomps):
+    """Check the keyword arguments for the decompose transform for the graph-based decomposition."""
+
+    if not qml.decomposition.enabled_graph() and (fixed_decomps or alt_decomps):
+        raise ValueError(
+            "The fixed_decomps and alt_decomps arguments must be used with the experimental graph-based decomposition."
+            "Please enable the decomposition graph with qml.decomposition.enable_graph()."
+        )
+
+
+def _get_decomp_graph(operations, target_gate_names, fixed_decomps, alt_decomps):
+    """Create and solve a DecompositionGraph instance to optimize the decomposition."""
+
+    # Exclude the gates in fixed_decomps and alt_decomps from the target gate set.
+    if fixed_decomps:
+        target_gate_names -= set(
+            op.name if isinstance(op, type) else op for op in fixed_decomps.keys()
+        )
+    if alt_decomps:
+        target_gate_names -= set(
+            op.name if isinstance(op, type) else op for op in alt_decomps.keys()
+        )
```

Comment:
Why is this necessary? If something is in the target gate set, I would expect that it isn't decomposed, regardless of whether decomposition rules are specified for it.